src/goto-instrument/contracts/assigns.cpp

```diff
@@ -52,6 +70,98 @@ static const slicet normalize_to_slice(const exprt &expr, const namespacet &ns)
   UNREACHABLE;
 }
 
+/// Result type for guarded assigns target pattern matching
+using if_match_resultt = optionalt<std::pair<exprt, exprt>>;
+
+/// Matches an expression with ID_NULL or  ID_constant and value "0",
+/// modulo some intermediate ID_typecasts
+bool match_NULL_or_zero(const exprt &expr)
+{
+  if(expr.id() == ID_typecast)
+    return match_NULL_or_zero(expr.operands().front());
+  else
+    return expr.id() == ID_NULL ||
+           (expr.id() == ID_constant &&
+            id2string(to_constant_expr(expr).get_value()) == "0");
+}
+
+/// Matches an expression of the form `guard ? target : (NULL|0)`
+/// modulo intermediary typecasts.
+/// \param expr the candidate expression
+/// \returns A pair `(guard,target)` if successful.
+if_match_resultt match_if(const exprt &expr)
+{
+  if(expr.id() == ID_typecast)
+  {
+    return match_if(expr.operands().front());
+  }
+  else if(expr.id() == ID_if)
+  {
+    if(!match_NULL_or_zero(expr.operands().at(2)))
+      return {};
+
+    return {{expr.operands().at(0), expr.operands().at(1)}};
+  }
+  else
+    return {};
+}
+
+/// Pattern match the given expression for a guarded pattern
+/// and builds guarded_slice expression as follows:
+///
+/// ```
+/// case expr of
+///  | guard ? target : NULL ->
+///          {guard,
+///           target,
+///           address_of{target},
+///           size_of_expr{target}}
+///  | __CPROVER_POINTER_OBJECT(guard ? target : NULL) ->
+///          {guard,
+///           __CPROVER_POINTER_OBJECT(target),
+///           address_of{target-offset(target)},
+///           object_size{target}}
+///  | other ->
+///          {true,
+///           expr,
+///           address_of{expr},
+///           object_size{expr}}
+/// ```
+static const guarded_slicet
+normalize_to_guarded_slice(const exprt &expr, const namespacet &ns)
+{
+  if(expr.id() == ID_pointer_object)
+  {
+    if_match_resultt match_opt = match_if(expr.operands().at(0));
+    if(match_opt.has_value())
+    {
+      const auto &match = match_opt.value();
+      const auto &new_expr = pointer_object(match.second);
+      const auto &slice = normalize_to_slice(new_expr, ns);
+      return {match.first, new_expr, slice.first, slice.second};
```

Comment:
It might be worth adding a constructor that takes a `slicet` to avoid this `slice.first, slice.second` (which repeats below).